### ISSUE TITLE
[Feature request] Support for smart cards

### ISSUE THREAD

**Issue by Eledril1** [adrienverge/openfortivpn#111]
I would like to know, if you plan to add support for Smart Cards, for example by OpenSC library.

**Comment by DimitriPapadopoulos** [comment]
@rmuehl Speaking for myself (not the other maintainers) I have no clue where to start from.
Do you have some documentation/links on what we can expect to find on smart cards? Perhaps a certificate?

**Comment by uschwarz** [comment]
If you go the pkcs11 route, openssh's ssh-agent source might be a good starting point, because that should in essence just forward stuff to the card's crypto chip, and it's not that huge, and the interesting stuff is ifdef'd.

**Comment by rmuehl** [comment]
On a smartcard you find the user-cert and the user-key. The key is protected, and gets unlocked by a PIN (U2F). So instead of using a cert.pem and a key.pem in the config there could be a PKCS11Provider to the opensc-pkcs11.so/dll.

You're using openssl so maybe this could be a good startingpoint:
https://github.com/OpenSC/pkcs11-helper
https://github.com/OpenSC/pkcs11-helper/blob/master/tests/test-openssl/test-openssl.c

**Comment by DimitriPapadopoulos** [comment]
Thank you for the examples. It really helps. I do understood the general direction but I just have a couple other questions:
* Out of curiosity, are these smart cards supported/documented by Fortinet or is this an independent way to provide the _cert.pem_ and _key.pem_ files expected by the Fortinet client/gateway? Just to understand the ecosystem, do you know any well-established providers of compatible smart cards + infrastructure to manage the smart cards?
* Am I correct that it is readily possible to extract the _cert.pem_ and _key.pem_ files from the smart card using some other software, and then feed these files to _openfortivpn_? I do realize this is **not** convenient! I only want to understand if and verify how the process is readily implemented. On which open source software would you rely to extract the _cert.pem_ and _key.pem_ files files from the smart card?

**Comment by uschwarz** [comment]
No, the smart card will never give out the key.pem again, which is what makes it a "something you own" factor instead of a "something you know" factor like key.pem. (I still think it's more useful to think of "something you own" as "something that can't be copied") 
In a very real workflow, the key is generated on the smart card itself and never leaves it, the card just offers sign, verify, encrypt, decrypt operations that use it. (And yes, there's management functions like "store this certificate along with the public key" and "give me the public key (so I can get a certificate)", but essentially the only management you can do with the secret key is to erase/replace it.)

**Comment by uschwarz** [comment]
Oh, and to answer your second question: the usual API in the middle is PKCS#11, which basically exposes the encrypt/decrypt/sign/verify functions of the card, so Fortinet server-side does not even see you're doing it. I think they sell (or used to) USB tokens, but those looked like rebranded Feitian ePass 2003, and Feitian's linux drivers are a bit rubbish in my experience. If you have a YubiKey, that can also work as a crypto smart card.

**Comment by rmuehl** [comment]
We are using Yubikey 4/Neo in our Company. Its cheap (~50 bucks) and easy to handle. Cuz its an USB Device, you don't need any smartcard-readers. And opensc can handle it well (along with may other smartcards).

In Win7/10 it works out of the box (with a windows mini driver). Forticlient can find the cert aslong it ist plugged in prior to start the client, and presents it in the dropdown. But there is no luck in mac/linux OSes.

Basically, opensc talks to the smartcard, and when the private key needs to be accessed, you need to enter a PIN to unlock. Its like a bankcard. And if you enter wrong PINs more then max_retries, it gets blocked. You can unblock the PIN with a PUK, if configured to, like a SIM card. Otherwise the Yubikey needs to be resettet und filled with new cert/key pair, PUK and PIN.

**Comment by DimitriPapadopoulos** [comment]
Last question, what happens if there are multiple smart cards? I suppose we need to select one of them. How do you distinguish between smart cards, probably boils down to how are are smart cards named?

I fear _openfortivpn_ is going to be less convenient than a GUI client which can list smart cards and let you select one of  them interactively.

**Comment by rmuehl** [comment]
opensc is enumerating multiple cards:

opensc-tool --list-readers 
# Detected readers (pcsc)
Nr.  Card  Features  Name
0    Yes             Yubico Yubikey NEO OTP+U2F+CCID 00 00
1    Yes             Yubico Yubikey 4 OTP+U2F+CCID 01 00

In this case its 2 Yubikeys, which are 2 readers ofcourse. I do not own a sc-reader with multiple card-slots, so I cant tell you how they are shown.... but I guess the numbers after CCID would look different.

So you could 

1. if you find only one card, use it
2. if you find more, list it on the command prompt (like otp) an let the user selext it by entering the number
3. you could invite a parameter in the config

**Comment by rmuehl** [comment]
This sounds interesting too:
https://github.com/adrienverge/openfortivpn/issues/111#issuecomment-422673676

It relies on libp11. This example is for a public_key auth, but it shows the procedure imho...

**Comment by rmuehl** [comment]
In the openssl manpage https://www.openssl.org/docs/man1.0.2/crypto/engine.html gives general information.

As far as I can guess, its getting loaded 
  char *engine_id = "pkcs11";
  tls_engine_load_dynamic_generic(pre_cmd, post_cmd, engine_id)

where the pre cmds are pretty much this (finally read from the config):
OpenSSL> engine -t dynamic -pre SO_PATH:/usr/lib/engines/engine_pkcs11.so
         -pre ID:pkcs11 -pre LIST_ADD:1 -pre LOAD 
         -pre MODULE_PATH:opensc-pkcs11.so

**Comment by rmuehl** [comment]
I managed to do a sslvpn connection with s_client using the Yubikey on the console. 
Tho I could not login (got no cookie), cuz I didn't look deeper into the html stuff, FortiGate diagnose showed that the client cert was sent and accepted.

`  // get the cert out from Yubikey (? reading cert in openssl is not implemented ?)
pkcs15-tool -r 1 > yubicert.pem


 //get the token-url
p11tool --list-token-urls


  // then connecting with s_client
openssl s_client -engine pkcs11 -keyform engine -key "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;serial=xxxxxxxxxx;token=myuser" -cert yubicert.pem -connect sslvpn.somwhere.tld
  // here you get asked for the PIN 
  
  // and when connected, do the html thing
POST /remote/logincheck HTTP/1.1
Host: sslvpn.somwhere.tld
User-Agent: Mozilla/5.0 SV1
Accept: text/plain
Accept-Encoding: identify
Content-Type: application/x-www-form-urlencoded
Cookie:
Content-Length: 90

username=myuser&credential=mypasswd&realm=&ajax=1&redir=%2Fremote%2Findex&just_logged_in=1`

Just to show how the workflow with openssl engine could be....

**Comment by aviadra** [comment]
I'd like this feature as well...

**Comment by rmuehl** [comment]
Did some testing and ugly hacks. Got it working:

> sudo ./openfortivpn 
**Enter PKCS#11 token PIN for myvpnuser:**
INFO:   Connected to gateway.
INFO:   Authenticated.
INFO:   Remote gateway has allocated a VPN.
INFO:   Got addresses: [10.10.10.10], ns [10.10.10.1, 10.10.10.2]
INFO:   Interface ppp0 is UP.
INFO:   Setting new routes...
INFO:   Tunnel is up and running.

Certificate and Key is read from Yubikey.

**Comment by rmuehl** [comment]
Some more detailed output:

> openfortivpn]$ sudo ./openfortivpn -v
DEBUG:  openfortivpn 1.7.1
DEBUG:  Loaded config file "/usr/local/etc/openfortivpn/config".                                                                                                       
DEBUG:  Config host = "sslvpn.somewhere.com"                                                                                                                                
DEBUG:  Config realm = ""                                                                                                                                              
DEBUG:  Config port = "443"                                                                                                                                            
DEBUG:  Config username =                                                                                                                                       
DEBUG:  Config password =                                                                                                                                    
DEBUG:  server_addr: <remote-ip>                                                                                                                                    
DEBUG:  server_port: 443                                                                                                                                               
DEBUG:  gateway_addr: <remote-ip>                                                                                                                                   
DEBUG:  gateway_port: 443                                                                                                                                              
Enter PKCS#11 token PIN for myvpnuser:
DEBUG:  Gateway certificate validation succeeded.
INFO:   Connected to gateway.
DEBUG:  Cookie: SVPNCOOKIE=EllMOWPMYxx2E2Q6O+2MQBxIXjE7xtPyzaX4x/UrdGt+1JnCJySLQ.......aRo/+aMpKj+QeWZPA7EQwfDpbJKO9iL60fIh2hiL/mjrWnCzoaXwCpADEgmnyMNYT1gg==                           
INFO:   Authenticated.
DEBUG:  Cookie: SVPNCOOKIE=EllMOWPMYxx2E2Q6O+2MQBxIXjE7xtPyzaX4x/UrdGt+1JnCJySLQ.......aRo/+aMpKj+QeWZPA7EQwfDpbJKO9iL60fIh2hiL/mjrWnCzoaXwCpADEgmnyMNYT1gg==                           
INFO:   Remote gateway has allocated a VPN.
DEBUG:  server_addr: <remote-ip>
DEBUG:  server_port: 443                                                                                                                                               
DEBUG:  gateway_addr: <remote-ip>                                                                                                                                   
DEBUG:  gateway_port: 443                                                                                                                                              
DEBUG:  Gateway certificate validation succeeded.                                                                                                                      
DEBUG:  pppd_read thread                                                                                                                                               
DEBUG:  ssl_read thread                                                                                                                                                
DEBUG:  ssl_write thread                                                                                                                                               
DEBUG:  if_config thread                                                                                                                                               
DEBUG:  pppd ---> gateway (16 bytes)                                                                                                                                   
DEBUG:  pppd_write thread                                                                                                                                              
DEBUG:  gateway ---> pppd (12 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (16 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (12 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (17 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (30 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (12 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (12 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (25 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (12 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (24 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (12 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (12 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (24 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (12 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (24 bytes)                                                                                                                                   
DEBUG:  pppd ---> gateway (12 bytes)                                                                                                                                   
DEBUG:  gateway ---> pppd (24 bytes)                                                                                                                                   
INFO:   Got addresses: [vpn-ip], ns [10.10.10.1, 10.10.10.2]
DEBUG:  Got Address: <vpn-ip>
DEBUG:  if_config: not ready yet...
DEBUG:  pppd ---> gateway (6 bytes)
DEBUG:  Got Address: <vpn-ip>
DEBUG:  Interface Name: ppp0
DEBUG:  Interface Addr: <vpn-ip>
INFO:   Interface ppp0 is UP.
INFO:   Setting new routes...
DEBUG:  ip route show to 0.0.0.0/0.0.0.0
DEBUG:  ip route show to <remoute-ip>/255.255.255.255
DEBUG:  Setting route to vpn server...
DEBUG:  ip route show to <remote-ip>/255.255.255.255 via 10.0.2.2 dev eth0
DEBUG:  ip route add to <remote-ip>/255.255.255.255 via 10.0.2.2 dev eth0
DEBUG:  Deleting the current default route...
DEBUG:  ip route del to 0.0.0.0/0.0.0.0 via 10.0.2.2 dev eth0
DEBUG:  Setting new default route...
DEBUG:  ip route add to 0.0.0.0/0.0.0.0 dev ppp0
INFO:   Tunnel is up and running.

**Comment by aviadrav** [comment]
@rmuehl, how can I reproduce your results?

**Comment by rmuehl** [comment]
It needs a lot of modifications in the sourcecode. I will streight up my ugly hacks, and then provide a patch or doing a pull request.
But I need some time for that....

**Comment by aviadra** [comment]
Know that I am literally on the edge of my seat for any code you have

**Comment by aviadra** [comment]
@rmuehl Please contact me via email so we can talk biz... The company I work at NEEDs what you have as you have it right **now**

**Comment by mrbaseman** [comment]
@rmuehl, could you share your code, please? Even if it's ugly as you say, it's probably a good starting point and we can improve it

**Comment by rmuehl** [comment]
Pull request created.

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman Martin, hasn't this issue been addressed?

**Comment by mrbaseman** [comment]
yes, that was #464 which has been merged into master

**Comment by mrbaseman** [comment]
This feature went into the [1.11.0 Release](https://github.com/adrienverge/openfortivpn/releases/tag/v1.11.0). 
Thanks again to @rmuehl for this great contribution.
